src/pydocstyle/parser.py

```diff
@@ -117,7 +118,25 @@ def is_public(self):
 
         This helps determine if it requires a docstring.
         """
-        return not self.name.startswith('_') or self.name.startswith('__')
+        module_name = Path(self.name).name
+        return (
+            not self._is_inside_private_package() and
+            self._is_public_name(module_name)
+        )
+
+    def _is_inside_private_package(self):
+        """Return True if the module is inside a private package."""
+        path = Path(self.name)
+        package_names = path.parts[:-1]
+        return any([self._is_private_name(package) for package in package_names])
+
+    def _is_public_name(self, module_name):
+        """Determine whether a "module name" (i.e. module or package name) is public."""
+        return not module_name.startswith('_') or module_name.startswith('__')
```

Comment:
I think we should modify it to start *and* end with double underscores for it to be private. thoughts?

Reply:
Sorry, can you elaborate on what you mean by this?

>  So something like `__main__.py` should be **public** but something like `__my_module.py` would hopefully be **private**.

If this is what you mean, then I'm re-working the code now. :+1: 

If you meant something else, like re-factoring the code to express the logic in a different way, then please provide code-snippets or more explanation.